### PR TITLE
snap-bootstrap: implement bind mounts for kernel, base

### PR DESCRIPTION
**Pull request by xnox**
Note this is backwards compatible version with the current initrd. And is effectively a no-op, as nothing new happens unless the sysroot*.mount units are dropped from the core-initrd project.

I need this functionality to become dynamic, such that the same core-initrd & snap-bootstrap can be used by either classic encrypted systems, or UC20 encrypted systems.

### CODE REVIEW

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 1]
```diff
@@ -178,6 +178,44 @@ func generateInitramfsMounts() (err error) {
 	return nil
 }
 
+func doBaseBindMount() (err error) {
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:181]: since these functions only do the mounts if the mount units don't exist, then this should be called

```suggestion
func maybeDoBaseBindMount() (err error) {
```

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 2]
```diff
@@ -178,6 +178,44 @@ func generateInitramfsMounts() (err error) {
...
+	}
+	if !osutil.FileExists(filepath.Join(dirs.GlobalRootDir, "/usr/lib/systemd/system/sysroot.mount")) {
+		if err := doSystemdMount(filepath.Join(boot.InitramfsRunMntDir, "base"), boot.InitramfsSysrootDir, bindSystemdOpts); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func doKernelBindMount() (err error) {
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:198]: same
```suggestion
func maybeDoKernelBindMount() (err error) {
```

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 3]
```diff
@@ -178,6 +178,44 @@ func generateInitramfsMounts() (err error) {
...
+func doBaseBindMount() (err error) {
+	// Bind-mount base snap to /sysroot as the start of pivot-root
+	// rootfs. Previously this was in the core-initrd, but it is
+	// best to be controlled by snap-bootstrap. NB! This needs to
+	// be backwards compatible, because older initrds are often
+	// repacked with newer snap-bootstrap.
+	bindSystemdOpts := &systemdMountOptions{
+		Bind: true,
+	}
+	if !osutil.FileExists(filepath.Join(dirs.GlobalRootDir, "/usr/lib/systemd/system/sysroot.mount")) {
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:190]: nice, we had it on our long todo list to do this, I like this method of doing it

File: cmd/snap-bootstrap/cmd_initramfs_mounts_test.go [hunk 4]
```diff
@@ -129,6 +129,18 @@ var (
...
+func makeOldCoreInitrdUnits(c *C) () {
+	// fake old core-initrd bind-mounts
+	systemPath := filepath.Join(dirs.GlobalRootDir, "/usr/lib/systemd/system")
+	os.MkdirAll(systemPath, os.ModePerm)
+	for _, unit := range []string{"sysroot.mount", "sysroot-usr-lib-modules.mount", "sysroot-usr-lib-firmware.mount"} {
+		f, err := os.Create(filepath.Join(systemPath, unit))
+		c.Assert(err, IsNil)
+		f.Close()
+	}
+
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts_test.go:141]: extra whitespace

```suggestion
```

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 5]
```diff
@@ -1538,6 +1576,14 @@ func generateMountsModeRun(mst *initramfsMountsState) error {
 		}
 	}
 
+	// Setup bind-mounts in /sysroot
+	if err := doBaseBindMount(); err != nil {
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1580]: this needs unit tests for the case where the mount unit files don't exist, let me know if you need help writing these tests, probably you wanna start with copying and modifying these 2 tests:
* `TestInitramfsMountsRunModeWithSaveHappyRealSystemdMount`
* `TestInitramfsMountsRunModeNoSaveUnencryptedHappy`

and have the start of those tests delete the mount unit files that were mocked in the `SetUpTest()`. You will then see some failures about the expected systemctl calls and probably just need to add these new calls in.

**Review by anonymouse64 — COMMENTED** [review]
Looks good to me, a couple nitpicks about naming and more unit tests we need here, but let me know if you need hep writing those unit tests since the cmd_initramfs_mounts_test.go file is quite unwieldy.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 5 above)
Comment (anonymouse64) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1580]: ah I see you added some tests in https://github.com/snapcore/snapd/pull/10351, typically we prefer to keep the tests in the same PR as the one that adds the features, so if you could just move those tests to this PR that would simplify reviewing this  PR for others. 

Also I see that in this PR you mock the mount unit files existing and don't modify the tests while the other PR has tests about when the mount units are missing - we need both kinds of tests hence my suggestion to add specific tests which check that things work when the mount units are missing, but since actually that situation will become "legacy" support, maybe we ought to flip it and modify all the tests the way you did in #10351, then take the tests I mentioned above and mock add the mount unit files and those two tests can be for the legacy situation. Does that make sense?

**Comment by pedronis** [comment]
Marked to run the nested tests, that will required closing/reopening time when things are deemed in shape for that

**Comment by anonymouse64** [comment]
@xnox we generally refrain from doing force pushes when there has been a review on the PR, otherwise we have to re-review the entire thing, whereas if you just push new commits or merge commits, GitHub can show us the diff of just the new stuff you pushed after the review was sent in.

**Comment by xnox** [comment]
> @xnox we generally refrain from doing force pushes when there has been a review on the PR, otherwise we have to re-review the entire thing, whereas if you just push new commits or merge commits, GitHub can show us the diff of just the new stuff you pushed after the review was sent in.

you asked me to rewrite tests..... as you said you didn't want these ones at all....

**Comment by anonymouse64** [comment]
@xnox yeah in this case we would usually just push a commit which deleted the tests we don't want and added the tests we do want

**Review by anonymouse64 — APPROVED** [review]
LGTM, the test cases look sufficient for this

**Comment by anonymouse64** [comment]
Also, due to a bug in GitHub Actions, we can't run the nested tests unless the PR is opened with the "run nested" label applied, so I'm going to close and re-open this PR so that we can get the nested test coverage too

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 6]
```diff
@@ -1538,6 +1576,14 @@ func generateMountsModeRun(mst *initramfsMountsState) error {
 		}
 	}
 
+	// Setup bind-mounts in /sysroot
+	if err := maybeDoBaseBindMount(); err != nil {
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1580]: @anonymouse64 @xnox I'm confused but if we do this only for run mode but to do something it requires the units not to be present, will this not break recover and install, or is the mounting to the tmpfs done by different units in that case?

**Review by pedronis — COMMENTED** [review]
I'm not quite sure what's the effect of this together with the changes it anticipates to the initramfs units for install and recover mode

**Review by pedronis — CHANGES_REQUESTED** [review]
this should probably repack initramfs in spread dropping those units to check that things works once they are gone

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 6 above)
Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1580]: In any UC20 mode (run / install / recover) sysroot.mount is a bind-mount of /run/mnt/ubuntu-data. Whereas in the classic mode (i.e. cloudimg-rootfs) it will be a direct mount of the rootfs.

Also in UC20 mode bind mounts of the kernel snap's modules & firmware must happen in the initrd. Whereas in classic mode there is no kernel snap, and modules/firmware is expected to be available in the rootfs already.

Currently the bind mounts are present as static units in the core-initrd, which make it impossible to make it usable for either UC20 or classic modes simultaniously.

This is similar change, to how in the past "ubuntu-data" was mounted with static units baked into the core-initrd, but then during early UC20 development got moved to be generated by snap-bootstrap. Which has allowed for more strict checks on which ubuntu-data is mounted.

**Comment by xnox** [comment]
> this should probably repack initramfs in spread dropping those units to check that things works once they are gone

Yes, very good suggestion. Also let me open a core-initrd pull request that drops these static bind mounts, referencing this PR.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 6 above)
Comment (anonymouse64) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1580]: Indeed this PR as it stands is insufficient, I am unable to boot a UC20 system in install or recover mode with the mount units removed from the initrd and this branch as-is. It just hangs in the initrd presumably because sysroot-writable has 
`After=run-mnt-base.mount` in it, and when we are running in install mode, `maybeDoBaseBindMount()` doesn't get called and so we are stuck forever in the initrd.

I _think_ it should be enough to add calls to these two functions in `generateMountsCommonInstallRecover()` after line 1277, the `for _, essentialSnap := range essSnaps {` loop, but I haven't successfully tested this yet.

I'm working on a spread test to test the new case where these files are absent from the initrd to see that it passes (I expect it won't ootb with this branch), and then in the future when the initrd is updated to remove these mount units, we can flip the tests to have this spread test _add_ those mount units to the initrd to confirm that it boots properly.

**Comment by anonymouse64** [comment]
@xnox so as mentioned I had a try at writing a spread test to check if this branch works on systems with initrds without the mount units present, just by deleting those mount units and the branch does not work. There seems to be some issue with populate-writable.service, in that it doesn't run when those mount unit files are not present in the initrd as the initrd just hangs after running through the-tool and has `sysroot.mount` and `sysroot-writable.mount` and friends setup, but populate-writable.service just never runs. 

So if you have a patch to core-initrd which changes the mount units (and possibly other things, not sure what else is needed to be changed) as needed in order to actually test this patch out, please let me know and I'll take a look and try to adapt that into a spread test we can run on this branch to demonstrate that it works for the new world as well as the old world, since it's clear this branch works for the old world, but it doesn't appear to work for the new world.

**Comment by xnox** [comment]
@anonymouse64 it's impossible to implement core-initrd changes because these snap-bootstrap changes are not merged.

it's not possible to write integration tests for this unmerged snap-bootstrap code; because core-initrd changes are not there.

Something needs to give, to unblock this deadlock. I cannot simultaniously change both core-initrd and snap-bootstrap.

**Comment by anonymouse64** [comment]
@xnox sorry if I was unclear, what I was trying to explain is that this branch as it currently exist does not work in the future case where you remove those mount units (contrary to what you claim in your comments here).

What needs to happen before this PR is mergable is that you need to provide us with details of how the initrd will be changed in the future to implement these changes so that we can write a spread test which demonstrates that these snap-bootstrap changes in this PR will work with those changes. Because otherwise we are merging code here when we don't know if the changes you plan to make to the initrd are sufficient.

Basically consider the case where we merge this and you then go try to enable core-initrd to work, and you find that actually snap-bootstrap needs to do more work, then the code that we merged here was wrong, but we merged it without any idea of how broken the world will be with this code because we don't know what world it will be run in.

**Comment by xnox** [comment]
ok, so i fork snap-bootstrap in core-initrd, change core-initrd upload that and stop using snapd-bootstrap from snap package, then if and when these changes are merged, switch back to snap-bootstrap that the package provides....

cause i can't land core-initrd changes whilst using old snap-bootstrap.

how am i supposed to provide core-initrd changes for you then?

**Comment by anonymouse64** [comment]
@xnox what I would suggest to do is the following:

1. build locally this branch of snap-bootstrap
2. put that snap-bootstrap into a local initrd
3. test with that local version of core-initrd and that local version of snap-bootstrap until such a time when it works as you describe
4. then create a diff between the core-initrd as it exists today on main and what you changed it to
5. provide us that diff, we will apply it here in this PR as part of a spread test
6. the spread test will run and it will demonstrate that snap-bootstrap will operate correctly in both the environment today on the main branch and also in the environment how you expect the initrd to change 
7. then we merge this code to snap-bootstrap
8. after some gears crank, then you can propose the changes to core-initrd that are effectively the patch we have in the spread test
9. if more changes are necessary then we start again at step 1 

This ensures that we never have untested code in snap-bootstrap. Does that make sense?

**Comment by xnox** [comment]
> This ensures that we never have untested code in snap-bootstrap. Does that make sense?

this would yield to forking unmerged core-initrd code, into unmerged snapd-code, to test unmerged snapd code...... or am i missing something?

**Comment by xnox** [comment]
I am concerned at doing any proof of concept work, which will be solely use to assess usefulness of a single pr, which would then be reverted and dropped straight away, without being used ever again.

**Comment by xnox** [comment]
UC team proposing to move these bind-mounts out of initrd and out of snap-bootstrap and into /etc/fstab inside the core22 snap instead.

Thus this PR will not be needed....... if snapd team unblocks UbuntuCore team to land removal of writable paths stuff.
